src/dex/nerve/nerve-math.ts

```diff
@@ -0,0 +1,317 @@
+    state: PoolState,
+    tokenIndexFrom: number,
+    tokenIndexTo: number,
+    dx: bigint,
+    blockTimestamp: bigint,
+  ) {
+    const xp = this._xp(state);
+
+    // uint256 x = dx.mul(self.tokenPrecisionMultipliers[tokenIndexFrom])
+    //    .add(xp[tokenIndexFrom]);
+    const x =
+      dx * state.tokenPrecisionMultipliers[tokenIndexFrom] + xp[tokenIndexFrom];
+
+    const y = this._getY(
+      state,
+      tokenIndexFrom,
+      tokenIndexTo,
+      x,
+      xp,
+      blockTimestamp,
+    );
+
+    if (y === null) {
+      state.isValid = false;
+      return null;
+    }
+
+    // dy = xp[tokenIndexTo].sub(y).sub(1);
+    let dy = xp[tokenIndexTo] - y - ONE;
+
+    // dyFee = dy.mul(self.swapFee).div(FEE_DENOMINATOR);
+    const dyFee = (dy * state.swapFee) / this.FEE_DENOMINATOR;
+
+    // dy = dy.sub(dyFee).div(self.tokenPrecisionMultipliers[tokenIndexTo]);
+    dy = (dy - dyFee) / state.tokenPrecisionMultipliers[tokenIndexTo];
+    return { dy, dyFee };
+  }
+
+  calculateWithdrawOneToken(
+    state: PoolState,
+    tokenAmount: bigint,
+    tokenIndex: number,
+    blockTimeStamp: bigint,
+  ) {
+    const out = this._calculateWithdrawOneTokenDY(
+      state,
+      tokenIndex,
+      tokenAmount,
+      blockTimeStamp,
+    );
+    if (out === null) return null;
+    let { dy, newY } = out;
+
+    // uint256 dySwapFee = _xp(self)[tokenIndex].sub(newY)
+    //  .div(self.tokenPrecisionMultipliers[tokenIndex]).sub(dy)
+    const dySwapFee =
+      (this._xp(state)[tokenIndex] - newY) /
+        state.tokenPrecisionMultipliers[tokenIndex] -
+      dy;
+
+    // dy = dy.mul(FEE_DENOMINATOR.sub(calculateCurrentWithdrawFee(self, account)))
+    //  .div(FEE_DENOMINATOR);
+    dy =
+      (dy * (this.FEE_DENOMINATOR - this._calculateCurrentWithdrawFee(state))) /
+      this.FEE_DENOMINATOR;
+
+    return { dy, dyFee: dySwapFee };
+  }
+
+  swapUnderlying(state: PoolState) {}
+
+  protected _getNumTokens(state: PoolState) {
+    return biginterify(state.tokenPrecisionMultipliers.length);
+  }
+
+  protected _calculateWithdrawOneTokenDY(
+    state: PoolState,
+    tokenIndex: number,
+    tokenAmount: bigint,
+    blockTimestamp: bigint,
+  ) {
+    const numTokens = this._getNumTokens(state);
+    const xp = this._xp(state);
+    const v = {
+      d0: ZERO,
+      d1: ZERO,
+      newY: ZERO,
+      feePerToken: ZERO,
+      preciseA: ZERO,
+    };
+    v.preciseA = this._getAPrecise(state, blockTimestamp);
+    const d0 = this._getD(state, xp, v.preciseA);
+    if (d0 === null) return null;
+    else v.d0 = d0;
+
+    // v.d1 = v.d0.sub(tokenAmount.mul(v.d0).div(self.lpToken.totalSupply()));
+    v.d1 = v.d0 - (tokenAmount * v.d0) / state.lpToken_supply;
+    const newY = this._getYD(state, v.preciseA, tokenIndex, xp, v.d1);
+    if (newY === null) return null;
+    else v.newY = newY;
+
+    const xpReduced: bigint[] = [];
+
+    v.feePerToken = this._feePerToken(state);
+    for (let i = 0; i < numTokens; i++) {
```

Comment:
Does this work? Isn't `numTokens` a bigint while i is a `number`

Reply:
this somehow magically works

Reply:
Pure lucky on my side that it actually works.

> A Number value and a BigInt value may be compared as usual:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt